File: WorkcraftCore/src/main/java/org/workcraft/dom/visual/connections/VisualConnectionGui.java
package org.workcraft.dom.visual.connections;

import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.Shape;
import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;

import org.workcraft.dependencymanager.advanced.core.Expression;
import org.workcraft.dom.visual.ColorisableGraphicalContent;
import org.workcraft.dom.visual.DrawHelper;
import org.workcraft.dom.visual.DrawRequest;
import org.workcraft.dom.visual.Touchable;
import org.workcraft.gui.Coloriser;
import org.workcraft.util.Function;
import org.workcraft.util.Function3;
import org.workcraft.util.Geometry;

public class VisualConnectionGui {

	public interface ExprConnectionGui {
		Expression<? extends Touchable> shape();
		Expression<? extends ColorisableGraphicalContent> graphicalContent();
		Expression<? extends ParametricCurve> parametricCurve();
	}

	public static Function<ParametricCurve, Touchable> connectionTouchableMaker = new Function<ParametricCurve, Touchable>(){
		@Override
		public Touchable apply(final ParametricCurve curve) {
			return makeConnectionTouchable(curve);
		}
	};

	public static Touchable makeConnectionTouchable(final ParametricCurve curve) {
		return new Touchable() {
			@Override
			public boolean hitTest(Point2D point) {
				return curve.getPointOnCurve(curve.getNearestPointT(point)).distance(point) < VisualConnection.HIT_THRESHOLD;
			}
			
			@Override
			public Point2D getCenter()
			{
				return curve.getPointOnCurve(0.5);
			}
			
			@Override
			public Rectangle2D getBoundingBox() {
				return curve.getBoundingBox();
			}
		};
	}
	
	public static Function3<PartialCurveInfo, VisualConnectionProperties, Shape, ColorisableGraphicalContent> connectionGraphicalContentMaker = new Function3<PartialCurveInfo, VisualConnectionProperties, Shape, ColorisableGraphicalContent>(){
		@Override
		public ColorisableGraphicalContent apply(final PartialCurveInfo cInfo, final VisualConnectionProperties connProps, final Shape connectionShape) {
			return makeGraphicalContent(cInfo, connProps, connectionShape);
		}
	};

	public static ColorisableGraphicalContent makeGraphicalContent(final PartialCurveInfo cInfo, final VisualConnectionProperties connProps, final Shape connectionShape) {
		return new ColorisableGraphicalContent() {
			@Override
			public void draw(DrawRequest r) {
				Graphics2D g = r.getGraphics();
				
				Color color = Coloriser.colorise(connProps.getDrawColor(), r.getColorisation().getColorisation());
				g.setColor(color);
				g.setStroke(connProps.getStroke());
				g.draw(connectionShape);
				
				if (connProps.hasArrow())
					DrawHelper.drawArrowHead(g, color, cInfo.arrowHeadPosition, cInfo.arrowOrientation, 
							connProps.getArrowLength(), connProps.getArrowWidth());
			}
		};
	}
	
	public static ConnectionGui getConnectionGui(final VisualConnectionProperties properties, final VisualConnectionContext context,  VisualConnectionData data) {
		final ParametricCurve curve = data.accept(new ConnectionDataVisitor<ParametricCurve>() {
			@Override
			public ParametricCurve visitPolyline(PolylineData polyline) {
				return PolylineGui.makeCurve(properties, context, polyline.controlPoints());
			}

			@Override
			public ParametricCurve visitBezier(BezierData bezier) {
				return BezierGui.makeCurve(properties, context, bezier.cp1(), bezier.cp2());
			}
		});
		PartialCurveInfo curveInfo = Geometry.buildConnectionCurveInfo(properties, context, curve, 0);
		Shape visiblePath = curve.getShape(curveInfo.tStart, curveInfo.tEnd);
		final ColorisableGraphicalContent gc =makeGraphicalContent(curveInfo, properties, visiblePath);
		final Touchable touchable = makeConnectionTouchable(curve);
		
		return new ConnectionGui() {
			@Override
			public Touchable shape() {
				return touchable;
			}
			
			@Override
			public ParametricCurve parametricCurve() {
				return curve;
			}
			
			@Override
			public ColorisableGraphicalContent graphicalContent() {
				return gc;
			}
		};
	}

	static <N> VisualConnectionGraphicalPropertiesImpl getConnectionProperties(final VisualConnection<N> connection) {
		return new VisualConnectionGraphicalPropertiesImpl(connection);
	}
}

File: ScalaGraphEditorUtil/src/main/scala/org/workcraft/dom/visual/connections/VisualConnectionGui.scala
package org.workcraft.dom.visual.connections

import java.awt.Color
import java.awt.Graphics2D
import java.awt.Shape
import java.awt.geom.Point2D
import java.awt.geom.Rectangle2D
import org.workcraft.graphics.ColorisableGraphicalContent
import org.workcraft.graphics.Touchable
import org.workcraft.scala.Expressions._
import org.workcraft.graphics.DrawRequest
import org.workcraft.graphics.Coloriser
import org.workcraft.graphics.TouchableC
import org.workcraft.graphics.BoundingBox
import org.workcraft.graphics.PartialCurveInfo
import org.workcraft.graphics.VisualCurveProperties
import java.awt.geom.Path2D
import java.awt.geom.AffineTransform
import java.awt.BasicStroke
import org.workcraft.graphics.Geometry.buildConnectionCurveInfo
import org.workcraft.graphics.BoundedColorisableGraphicalContent
import org.workcraft.graphics.GraphicalContent
import org.workcraft.graphics.Java2DDecoration._
import org.workcraft.graphics.Colorisation

object VisualConnectionGui {

  val HitThreshold = 0.2

  def drawArrowHead(g: Graphics2D, color: Color, headPosition: Point2D, orientation: Double, length: Double, width: Double) = {
    val arrowShape = new Path2D.Double()
    arrowShape.moveTo(-length, -width / 2)
    arrowShape.lineTo(-length, width / 2)
    arrowShape.lineTo(0, 0)
    arrowShape.closePath()

    val arrowTransform = new AffineTransform()
    arrowTransform.translate(headPosition.getX(), headPosition.getY())
    arrowTransform.rotate(orientation)

    val transformedArrowShape = arrowTransform.createTransformedShape(arrowShape)

    g.setColor(color)
    g.setStroke(new BasicStroke(width.toFloat))
    g.fill(transformedArrowShape)
  }

  def arrowHead(color: Color, headPosition: Point2D, orientation: Double, length: Double, width: Double) = {
    val arrowShape = new Path2D.Double()
    arrowShape.moveTo(-length, -width / 2)
    arrowShape.lineTo(-length, width / 2)
    arrowShape.lineTo(0, 0)
    arrowShape.closePath()

    val arrowTransform = new AffineTransform()
    arrowTransform.translate(headPosition.getX(), headPosition.getY())
    arrowTransform.rotate(orientation)

    val transformedArrowShape = arrowTransform.createTransformedShape(arrowShape)

    BoundedColorisableGraphicalContent(
      ColorisableGraphicalContent(colorisation => GraphicalContent(g => { g.setColor(Coloriser.colorise(color, colorisation.foreground)); g.setStroke(new BasicStroke(width.toFloat)); g.fill(transformedArrowShape) })),
      BoundingBox(transformedArrowShape.getBounds2D()))
  }

  case class ExprConnectionGui(
    shape: Expression[Touchable], graphicalContent: Expression[ColorisableGraphicalContent], parametricCurve: Expression[org.workcraft.graphics.ParametricCurve])

  def makeConnectionTouchable(curve: org.workcraft.graphics.ParametricCurve, partial: PartialCurveInfo): TouchableC =
    new TouchableC(new Touchable {
      override def hitTest(point: Point2D.Double) = {
        val nearestT = curve.nearestPointT(point)
        nearestT < partial.tEnd && nearestT > partial.tStart && (curve.pointOnCurve(nearestT).distance(point) < HitThreshold)
      }
      override def boundingBox = BoundingBox(curve.boundingBox)
    }, curve.pointOnCurve(0.5))

  def makeGraphicalContent(cInfo: PartialCurveInfo, curve: org.workcraft.graphics.ParametricCurve, connProps: VisualCurveProperties, connectionShape: Shape) = {
    new ColorisableGraphicalContent {
      override def draw(r: DrawRequest) {
        val g = r.graphics

        val color = Coloriser.colorise(connProps.color, r.colorisation.foreground)
        g.setColor(color)
        g.setStroke(connProps.stroke)
        g.draw(connectionShape)

        connProps.arrow.foreach(arrow =>
          drawArrowHead(g, color, cInfo.arrowHeadPosition, cInfo.arrowOrientation,
            arrow.length, arrow.width))

	connProps.label.foreach ( label => {
	  val p = curve.pointOnCurve(0.5)
	  val d = curve.derivativeAt(0.5)
	  val dd = curve.secondDerivativeAt(0.5)

	  val q = if (d.getX < 0)  (d * -1)  else d

	  val labelPosition = new Point2D.Double(label.bounds.logical.getCenterX, 
						 if ((q cross dd) > -0.01) label.bounds.logical.getMaxY
						 else label.bounds.logical.getMinY)

	  val offset = p - labelPosition

	  val transform = AffineTransform.getTranslateInstance(offset.getX, offset.getY)
	  transform.concatenate(AffineTransform.getRotateInstance(q.getX, q.getY, labelPosition.getX, labelPosition.getY))

	  label.transform(transform).cgc.draw(r)
	})
      }
    }
  }

  def getConnectionGui(properties: VisualCurveProperties, context: VisualConnectionContext, data: StaticVisualConnectionData) = {
    val curve = data match {
      case Polyline(cps) => PolylineGui.makeCurve(properties, context, cps)
      case Bezier(cp1, cp2) => BezierGui.makeCurve(context, cp1, cp2)
    }
    val curveInfo = buildConnectionCurveInfo(properties.arrow, context.c1.touchable, context.c2.touchable, curve, 0)
    val visiblePath = curve.shape(curveInfo.tStart, curveInfo.tEnd)
    val gc = makeGraphicalContent(curveInfo, curve, properties, visiblePath)
    val touchable = makeConnectionTouchable(curve, curveInfo)

    new ConnectionGui(touchable, gc, curve)
  }
}
